File: docker/hik-sdk/Samples/32/Java/ConnectSpecCamera/ConnectSpecCamera.java
/***************************************************************************************************
 * @file      ConnectSpecCamera.java
 * @breif     Use functions provided in MvCameraControlWrapper.jar to grab images
 * @author    hulongcheng
 * @date      2021/07/12
 *
 * @warning
 * @version   V1.0.0  2021/07/12 Create this file
 * @since     2021/07/14
 **************************************************************************************************/


import MvCameraControlWrapper.CameraControlException;
import MvCameraControlWrapper.MvCameraControl;
import MvCameraControlWrapper.MvCameraControlDefines;

import java.util.ArrayList;
import java.util.NoSuchElementException;
import java.util.Scanner;

import static MvCameraControlWrapper.MvCameraControl.*;
import static MvCameraControlWrapper.MvCameraControlDefines.*;

public class ConnectSpecCamera {

 private static void printFrameInfo(MV_FRAME_OUT_INFO stFrameInfo) {
        if (null == stFrameInfo) {
            System.err.println("stFrameInfo is null");
            return;
        }

        StringBuilder frameInfo = new StringBuilder("");
        frameInfo.append(("\tFrameNum[" + stFrameInfo.frameNum + "]"));
        frameInfo.append("\tWidth[" + stFrameInfo.width + "]");
        frameInfo.append("\tHeight[" + stFrameInfo.height + "]");
        frameInfo.append(String.format("\tPixelType[%#x]", stFrameInfo.pixelType.getnValue()));

        System.out.println(frameInfo.toString());
    }
    static class GetOneFrameThread extends Thread {

        private Handle mHandle;
        private boolean flag = true;

        public GetOneFrameThread(Handle handle) {
            this.mHandle = handle;
        }

        public void stopThread() {
            flag = false;
        }

        @Override
        public void run() {
            super.run();

            // Get payload size
            MVCC_INTVALUE stParam = new MVCC_INTVALUE();
            int nRet = MvCameraControl.MV_CC_GetIntValue(mHandle, "PayloadSize", stParam);
            if (MV_OK != nRet) {
                System.err.printf("Get PayloadSize fail, errcode: [%#x]\n", nRet);
                return;
            }

            while (flag) {
                MV_FRAME_OUT_INFO stImageInfo = new MV_FRAME_OUT_INFO();
                byte[] pData = new byte[(int)stParam.curValue];
                nRet = MvCameraControl.MV_CC_GetOneFrameTimeout(mHandle, pData, stImageInfo, 1000);
                if (MV_OK != nRet) {
                    System.err.printf("GetOneFrameTimeout fail, errcode:[%#x]\n", nRet);
                    break;
                } else {
                    printFrameInfo(stImageInfo);
                }
            }
        }
    }
	
	 public static boolean isNumeric(String strNum) {
        if (strNum == null) {
            return false;
        }
        try {
            int d = Integer.parseInt(strNum);
        } catch (NumberFormatException nfe) {
            return false;
        }
        return true;
    }

    public static void main(String[] args) {
        int nRet = MV_OK;
        int camIndex = -1;
        Handle hCamera = null;
      
        Scanner scanner = new Scanner(System.in);

		MV_CC_DEVICE_INFO stDevInfo = new MV_CC_DEVICE_INFO();
	
		// The camera IP that needs to be connected (based on actual padding)
		System.out.print("Please input Current Camera Ip :");
		String strCurrentIp = scanner.nextLine(); 
		
	
		
		 int nIp1, nIp2, nIp3, nIp4, nIp;

	    String[] parts = strCurrentIp.split("\\."); 
 
        
        if (parts.length == 4 && isNumeric(parts[0]) && isNumeric(parts[1]) && isNumeric(parts[2]) &&(isNumeric(parts[3]))) {
             nIp1 = Integer.parseInt(parts[0]);
             nIp2 = Integer.parseInt(parts[1]);
			 nIp3 = Integer.parseInt(parts[2]);
			 nIp4 = Integer.parseInt(parts[3]);
        } else {
            System.out.println("Input currentIP Invalid");
             return;
        }

	
   	// The pc IP that needs to be connected (based on actual padding)
		System.out.print("Please input Net Export Ip : ");
		String strNetExport = scanner.nextLine(); 
		String[] Netparts = strNetExport.split("\\.");
		if (Netparts.length == 4 && isNumeric(Netparts[0]) && isNumeric(Netparts[1]) && isNumeric(Netparts[2]) &&(isNumeric(Netparts[3]))) {
             nIp1 = Integer.parseInt(Netparts[0]);
             nIp2 = Integer.parseInt(Netparts[1]);
			 nIp3 = Integer.parseInt(Netparts[2]);
			 nIp4 = Integer.parseInt(Netparts[3]);
        } else {
            System.out.println("Input NetIP Invalid");
            return;
        }

	
		stDevInfo.transportLayerType = MV_GIGE_DEVICE;// ch:仅支持GigE相机 | en:Only support GigE camera
		stDevInfo.gigEInfo.currentIp  = strCurrentIp;
		stDevInfo.gigEInfo.netExport = strNetExport;
	
	
        do {
            System.out.println("SDK Version " + MvCameraControl.MV_CC_GetSDKVersion());

			// Initialize SDK
		    nRet = MvCameraControl.MV_CC_Initialize();
		    if (MV_OK != nRet)
		    {
			   System.err.printf("Initialize SDK fail! nRet [0x%x]\n\n",nRet);
               break;
		    }


            // Create handle
            try {
                hCamera = MvCameraControl.MV_CC_CreateHandle(stDevInfo);
            }
            catch (CameraControlException e) {
                System.err.println("Create handle failed!" + e.toString());
                e.printStackTrace();
                hCamera = null;
                break;
            }

            // Open device
            nRet = MvCameraControl.MV_CC_OpenDevice(hCamera);
            if (MV_OK != nRet) {
                System.err.printf("Connect to camera failed, errcode: [%#x]\n", nRet);
                break;
            }
			
			

            // Make sure that trigger mode is off
            nRet = MvCameraControl.MV_CC_SetEnumValueByString(hCamera, "TriggerMode", "Off");
            if (MV_OK != nRet) {
                System.err.printf("SetTriggerMode failed, errcode: [%#x]\n", nRet);
                break;
            }

            // Start grabbing
            nRet = MvCameraControl.MV_CC_StartGrabbing(hCamera);
            if (MV_OK != nRet) {
                System.err.printf("Start Grabbing fail, errcode: [%#x]\n", nRet);
                break;
            }

            // Get frame
            GetOneFrameThread getOneFrameThread = new GetOneFrameThread(hCamera);
            getOneFrameThread.start();

            // Press enter and enter any characters to end the get frame thread
			String input = scanner.nextLine();
            while (scanner.hasNextLine()) {
                getOneFrameThread.stopThread();
                try {
                    getOneFrameThread.join();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                getOneFrameThread = null;
                break;
            }
            // Stop grabbing
            nRet = MvCameraControl.MV_CC_StopGrabbing(hCamera);
            if (MV_OK != nRet)
            {
                System.err.printf("StopGrabbing fail, errcode: [%#x]\n", nRet);
                break;
            }
        } while (false);

        if (null != hCamera)
        {
            // Destroy handle
            nRet = MvCameraControl.MV_CC_DestroyHandle(hCamera);
            if (MV_OK != nRet) {
                System.err.printf("DestroyHandle failed, errcode: [%#x]\n", nRet);
            }
        }
        MvCameraControl.MV_CC_Finalize();
        scanner.close();
    }

}

File: docker/hik-sdk/Samples/64/C++/General/ConnectSpecCamera/ConnectSpecCamera.cpp
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <pthread.h>
#include "MvCameraControl.h"

bool g_bExit = false;

// 等待用户输入enter键来结束取流或结束程序
// wait for user to input enter to stop grabbing or end the sample program
void PressEnterToExit(void)
{
    int c;
    while ( (c = getchar()) != '\n' && c != EOF );
    fprintf( stderr, "\nPress enter to exit.\n");
    while( getchar() != '\n');
    g_bExit = true;
    sleep(1);
}

static void* WorkThread(void* pUser)
{
    int nRet = MV_OK;

    MV_FRAME_OUT stImageInfo = {0};
    memset(&stImageInfo, 0, sizeof(MV_FRAME_OUT));

    while(1)
    {
        if(g_bExit)
        {
            break;
        }

        nRet = MV_CC_GetImageBuffer(pUser, &stImageInfo, 1000);
        if (nRet == MV_OK)
        {
            printf("Get One Frame: Width[%d], Height[%d], nFrameNum[%d]\n", 
                stImageInfo.stFrameInfo.nExtendWidth, stImageInfo.stFrameInfo.nExtendHeight, stImageInfo.stFrameInfo.nFrameNum);

            MV_CC_FreeImageBuffer(pUser, &stImageInfo);
        }
        else
        {
            printf("No data[0x%x]\n", nRet);
            break;        
        }
    }

    return 0;
}

int main()
{
    int nRet = MV_OK;
    void* handle = NULL;
    MV_CC_DEVICE_INFO stDevInfo = {0};
    MV_GIGE_DEVICE_INFO stGigEDev = {0};

    // ch:需要连接的相机ip(根据实际填充) | en:The camera IP that needs to be connected (based on actual padding)
    printf("Please input Current Camera Ip : ");
    char nCurrentIp[128];
    scanf("%s", (char*)&nCurrentIp);
    // ch:相机对应的网卡ip(根据实际填充) | en:The pc IP that needs to be connected (based on actual padding)
    printf("Please input Net Export Ip : ");
    char nNetExport[128];
    scanf("%s", (char*)&nNetExport);
    unsigned int nIp1, nIp2, nIp3, nIp4, nIp;

    sscanf(nCurrentIp, "%d.%d.%d.%d", &nIp1, &nIp2, &nIp3, &nIp4);
    nIp = (nIp1 << 24) | (nIp2 << 16) | (nIp3 << 8) | nIp4;
    stGigEDev.nCurrentIp = nIp;

    sscanf(nNetExport, "%d.%d.%d.%d", &nIp1, &nIp2, &nIp3, &nIp4);
    nIp = (nIp1 << 24) | (nIp2 << 16) | (nIp3 << 8) | nIp4;
    stGigEDev.nNetExport = nIp;

    stDevInfo.nTLayerType = MV_GIGE_DEVICE;// ch:仅支持GigE相机 | en:Only support GigE camera
    stDevInfo.SpecialInfo.stGigEInfo = stGigEDev;

    do 
    {
        // ch:初始化SDK | en:Initialize SDK
        nRet = MV_CC_Initialize();
        if (MV_OK != nRet)
        {
            printf("Initialize SDK fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:选择设备并创建句柄 | en:Select device and create handle
        nRet = MV_CC_CreateHandle(&handle, &stDevInfo);
        if (MV_OK != nRet)
        {
            printf("Create Handle fail! nRet[0x%x]\n", nRet);
            break;
        }

        // ch:打开设备 | en:Open device
        nRet = MV_CC_OpenDevice(handle);
        if (MV_OK != nRet)
        {
            printf("Open Device fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
        int nPacketSize = MV_CC_GetOptimalPacketSize(handle);
        if (nPacketSize > 0)
        {
            nRet = MV_CC_SetIntValueEx(handle,"GevSCPSPacketSize",nPacketSize);
            if(nRet != MV_OK)
            {
                printf("Warning: Set Packet Size fail nRet [0x%x]!\n", nRet);
            }
        }
        else
        {
            printf("Warning: Get Packet Size fail nRet [0x%x]!\n", nPacketSize);
        }

        
        // ch:设置触发模式为off | en:Set trigger mode as off
        nRet = MV_CC_SetEnumValue(handle, "TriggerMode", MV_TRIGGER_MODE_OFF);
        if (MV_OK != nRet)
        {
            printf("Set Trigger Mode fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:开始取流 | en:Start grab image
        nRet = MV_CC_StartGrabbing(handle);
        if (MV_OK != nRet)
        {
            printf("Start Grabbing fail! nRet [0x%x]\n", nRet);
            break;
        }

        pthread_t nThreadID;
        nRet = pthread_create(&nThreadID, NULL ,WorkThread , handle);
        if (nRet != 0)
        {
            printf("thread create failed.ret = %d\n",nRet);
            break;
        }

        PressEnterToExit();  

        // ch:停止取流 | en:Stop grab image
        nRet = MV_CC_StopGrabbing(handle);
        if (MV_OK != nRet)
        {
            printf("Stop Grabbing fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:关闭设备 | en:Close device
        nRet = MV_CC_CloseDevice(handle);
        if (MV_OK != nRet)
        {
            printf("Close Device fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:销毁句柄 | en:Destroy handle
        nRet = MV_CC_DestroyHandle(handle);
        if (MV_OK != nRet)
        {
            printf("Destroy Handle fail! nRet [0x%x]\n", nRet);
            break;
        }
        handle = NULL;
    } while (0);
  
    if (handle != NULL)
    {
        MV_CC_DestroyHandle(handle);
        handle = NULL;
    }
    

    // ch:反初始化SDK | en:Finalize SDK
    MV_CC_Finalize();

    printf("exit.\n");

    return 0;
}
